File: 152.maximum-product-subarray.py
#
# @lc app=leetcode id=152 lang=python3
#
# [152] Maximum Product Subarray
#

# Time:  O(n)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        global_max, local_max, local_min = float("-inf"), 1, 1
        for num in nums:
            local_max, local_min = max(num, local_max * num, local_min * num), min(
                num, local_max * num, local_min * num
            )
            global_max = max(global_max, local_max)
        return global_max


# @lc code=end

File: 152.maximum-product-subarray.js
/*
 * @lc app=leetcode id=152 lang=javascript
 *
 * [152] Maximum Product Subarray
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function (nums) {
  let res = -Infinity;
  let mx = 1;
  let mn = 1;
  for (const num of nums) {
    [mx, mn] = [
      Math.max(num, mx * num, mn * num),
      Math.min(num, mx * num, mn * num),
    ];
    res = Math.max(res, mx);
  }
  return res;
};
// @lc code=end
